packages/@react-native-windows/codegen/Document.md

```diff
@@ -0,0 +1,386 @@
+
+By calling `MakeModuleProvider<MyModule>`,
+if there is `using ModuleSpec = MyModuleSpec;` in `MyModule`,
+the compiler will compare `MyModule` with `MyModuleSpec` and see if all required members are defined.
+Failed to define any required member results in compiler errors telling you what member is needed.
+All acceptable member declarations in C++ for a missing member will be included in error messages,
+you could pick one and copy to your `MyModule`.
+
+`MakeTurboModuleProvider<MyModule>` does exactly what `MakeModuleProvider<MyModule>` does,
+but it requires `MyModule::ModuleSpec` to exist,
+unlike `MakeModuleProvider<MyModule>` where `MyModule::ModuleSpec` is optional.
+
+**Tips**: You could use `React::` if you think `winrt::Microsoft::ReactNative::` is too long after `#include <NativeModules.h>`.
+
+### REACT_MODULE
+
+The syntax of `REACT_MODULE` is `REACT_MODULE(moduleStruct, [optional]moduleName, [optional]eventEmitterName)`.
+
+- `moduleStruct`: the struct name that implements the TurboModule without namespace.
+- `moduleName`: the module name visible to JavaScript. The default value is `L"moduleStruct"`.
+- `eventEmitterName`: the default event emitter name. The default value is `L""`, which is `RCTDeviceEventEmitter`.
+
+### Members
+
+Any member in the struct implementing TurboModule functions should be attached with an attribute like `REACT_MODULE`.
+Here is the list of all available attributes:
+
+- `REACT_INIT`
+- `REACT_SYNC_METHOD`
+- `REACT_METHOD`
+- `REACT_GET_CONSTANTS`
+- `REACT_CONSTANT_PROVIDER`
+- `REACT_CONSTANT`
+- `REACT_EVENT`
+- `REACT_FUNCTION`
+
+### REACT_INIT
+
+`REACT_INIT` attaches to a function.
+Such function is called before a TurboModule is used in JavaScript,
+there is no deterministic timing about when this function will be called,
+and also no guarantee about in which thread this function will be called.
+
+The syntax of `REACT_INIT` is `REACT_INIT(method)`.
+
+- `method`: the C++ method to call to initialize this TurboModule.
+```c++
+REACT_INIT(Initialize)
+void Initialize(winrt::Microsoft::ReactNative::ReactContext const& reactContext) noexcept;
+```
+
+### REACT_SYNC_METHOD
+
+`REACT_SYNC_METHOD` attaches to a function.
+Such function is exposed in the TurboModule to JavaScript.
+There is no guarantee about in which thread this function will be called.
+
+The syntax of `REACT_SYNC_METHOD` is `REACT_SYNC_METHOD(method, [optional]methodName)`
+
+- `method`: the C++ method to call.
+- `methodName`: the method name that is visible to JavaScript. The default value is `L"method"`.
+
+```c++
+REACT_SYNC_METHOD(AddSync)
+int AddSync(int x, int y) noexcept
+{
+    return x + y;
+}
+
+REACT_SYNC_METHOD(AddSyncStatic)
+static int AddSync(int x, int y) noexcept
+{
+    return x + y;
+}
+```
+
+### REACT_METHOD
+
+`REACT_METHOD` is an async version of `REACT_SYNC_METHOD`.
+The attached function must returns `void`
+and its last argument must be `winrt::Microsoft::ReactNative::ReactPromise<T> const& result`.
```

Comment:
>last argument must be [](http://example.com/codeflow?start=8&length=21)

Technically, the NativeModules 2.0 offer different ways to return result from async methods. I wonder if the Promise-only is a requirement from the TurboModule specs?

Reply:
Actually TurboModule also supports this, I missed this part in the spec. The only difference is that, in JS side TurboModule async functions can only use Promise, but in C++ side even if you use callbacks, it will be converted to a Promise at runtime..